src/psbt.c

```diff
@@ -194,6 +194,25 @@ static const struct wally_tx_output *utxo_from_input(const struct wally_psbt *ps
                                      child_path, child_path_len); \
     }
 
+/* Add a taproot keypath to parent structs keypaths member */
+#define ADD_TAP_KEYPATH(PARENT) \
+    int PARENT ## _tap_keypath_add(struct PARENT *parent, \
+                               const unsigned char *pub_key, size_t pub_key_len, \
+                               const unsigned char *tapleaf_hashes, size_t tapleaf_hashes_len, \
+                               const unsigned char *fingerprint, size_t fingerprint_len, \
+                               const uint32_t *child_path, size_t child_path_len) { \
+        int ret; \
+        if (!parent) return WALLY_EINVAL; \
+        ret = wally_map_merkle_path_add(&parent->taproot_leaf_hashes, \
```

Comment:
so to be strictly correct here you need to validate the args for the second call before making the first, to avoid adding to only one map (where possible, operations should be atomic). You can do this with the validation function from `parent->taproot_leaf_paths`.